lua/doc/lua.1
.\" lua.man,v 1.7 2002/12/13 11:45:12 lhf Exp
.TH LUA 1 "2002/12/13 11:45:12"
.SH NAME
lua \- Lua interpreter
.SH SYNOPSIS
.B lua
[
.I options
]
[
.I script
[
.I args
]
]
.SH DESCRIPTION
.B lua
is the stand-alone Lua interpreter.
It loads and executes Lua programs,
either in textual source form or
in precompiled binary form.
(Precompiled binaries are output by
.BR luac ,
the Lua compiler.)
.B lua
can be used as a batch interpreter and also interactively.
.LP
The given
.I options
(see below)
are executed and then
the Lua program in file
.I script
is loaded and executed.
The given
.I args
are available to
.I script
as strings in a global table named
.BR arg .
If these arguments contain spaces or other characters special to the shell,
then they should be quoted
(but note that the quotes will be removed by the shell).
The arguments in
.B arg
start at 0,
which contains the string
.RI ` script '.
The index of the last argument is stored in
.BR "arg.n" .
The arguments given in the command line before
.IR script ,
including the name of the interpreter,
are available in negative indices in
.BR arg .
.LP
At the very start,
before even handling the command line,
.B lua
executes the contents of the environment variable
.BR LUA_INIT ,
if it is defined.
If the value of
.B LUA_INIT
is of the form
.RI `@ filename ',
then
.I filename
is executed.
Otherwise, the string is assumed to be a Lua statement and is executed.
.LP
Options start with
.B \-
and are described below.
You can use
.B "\--"
to signal the end of options.
.LP
If no arguments are given,
then
.B "\-v \-i"
is assumed when the standard input is a terminal;
otherwise,
.B "\-"
is assumed.
Interactive mode is explained in detail below.
.SH OPTIONS
.TP
.B \-
load and execute the standard input as a file,
that is,
not interactively,
even when the standard input is a terminal.
.TP
.BI \-e " stat"
execute statement
.IR stat .
You need to quote
.I stat 
if it contains spaces, quotes,
or other characters special to the shell.
.TP
.B \-i
enter interactive mode after
.I script
is executed.
In this mode,
.B lua
prompts the user,
reads lines from the standard input,
and executes them as they are read.
If a line does not contain a complete statement,
then a secondary prompt is displayed and
lines are read until a complete statement is formed or
a syntax error is found.
So, one way to interrupt the reading of an incomplete statement is
to force a syntax error:
adding a
.B `;' 
in the middle of a statement is a sure way of forcing a syntax error
(except inside multiline strings and comments; these must be closed explicitly).
If a line starts with
.BR `=' ,
then
.B lua
displays the values of all the expressions in the remainder of the
line. The expressions must be separated by commas.
The primary prompt is the value of the global variable
.BR _PROMPT ,
if this value is a string;
otherwise, the default prompt is used.
Similarly, the secondary prompt is the value of the global variable
.BR _PROMPT2 .
So,
to change the prompts,
set the corresponding variable to a string of your choice.
You can do that after calling the interpreter
or on the command line with
.BR "_PROMPT" "=\'lua: \'" ,
for example.
(Note the need for quotes, because the string contains a space.)
The default prompts are ``> '' and ``>> ''.
.TP
.BI \-l " file"
call
.BI require( file )
before executing
.IR script.
Typically used to load libraries
(hence the letter
.IR l ).
.TP
.B \-v
show version information.
.SH "SEE ALSO"
.BR luac (1)
.br
http://www.lua.org/
.SH DIAGNOSTICS
Error messages should be self explanatory.
.SH AUTHORS
R. Ierusalimschy,
L. H. de Figueiredo,
and
W. Celes
(lua@tecgraf.puc-rio.br)
.\" EOF


ChangeLog
2002-01-23  Peter Wang  <tjaden@users.sourceforge.net>

	* mylua.c: Seed lua random() function.

	* gamesrv.c: Changed the test for a very small yv from
	0 <= yv < 0.0000001 to 0 <= yv < 0.0001.

	* basic-player.lua: Added rest of death animations.

2002-01-22  Peter Wang  <tjaden@users.sourceforge.net>

	* basic-powerup.lua, basic-player.lua:
	Made lightamp powerup give player a big blue light.

	* basic-player.lua: Made local player's light turn red if gotten
	shot, and made corpses use the red light.

	* basic-player.lua:
	Added start of backpacks.  Also minor clean up for walking anim.

	* editor-init.lua:
	Some functions cannot be run while in the editor or it will crash, so
	we make them into dummy functions.
	
	* init.lua, editor/editor.c:
	Only run editor-init.lua when actually running the editor.
	This allows us to put in editor-specific code in there.

	* gamesrv.c, gameclt.c, basic-weapon.lua, init.lua:
	Added object type aliases.

	* gamesrv.c: Added network traffic monitoring.
	* textface.c, gamesrv.h: Added set_status() method to
	game_server_interface.

	* basic-player.lua, basic-powerup.lua:
	Made health items give health.

	* basic-weapon.lua, basic-powerup.lua, basic-base.lua:
	Allow a second argument to Objtype and Weapon, which should
	contain a table of fields to add into or override in the first
	argument.

	* basic-weapon.lua, basic-player.lua: Animated weapon firing.

	* netmsg.h, gamesrv.c, gameclt.c, basic-weapon.lua,
	* basic-player.lua: Added weapon switching with keyboard.

	* basic-weapon.lua, basic-player.lua: Added ammo limits.

2002-01-21  Peter Wang  <tjaden@users.sourceforge.net>

	* basic-player.lua: Made player bleed when on low health.

	* gameclt.h, bindgen.lua, basic-weapon.lua, basic-player.lua,
	* camera.h, camera.c, gameclt.c:
	Made switching to the sniper rifle increase the max distance the
	camera can see.

	* basic-weapon.lua: Changed the generalised declarators to take
	table arguments, which are more verbose and easier to read.

	* basic-player.lua, basic-weapon.lua:
	Use the right image for the weapon we are holding (not animated yet).

	* gameclt.c, bindgen.lua, gamesrv.h, gamesrv.c, netmsg.h:
	Added mechanism for server scripts to tell clients to execute object
	methods (i.e. remote calls).

	* object.h, object.c:
	Allowed object_call() to take an nargs argument.  The arguments should
	be already be pushed on the lua stack.

	* basic-base.lua: Added merge function.

	* gamesrv.c: Fixed a bug where sometimes you couldn't jump because
	the player's yv was very, very, very close to zero but not zero.

2002-01-20  Peter Wang  <tjaden@users.sourceforge.net>

	* gameclt.c, camera.h, camera.c:
	Added ability to "push" the camera, i.e. for the sniper rifle so you
	can see further that your mouse would normally allow on the screen.
	This isn't actually enabled when you pick up a sniper rifle, yet.

	* basic-weapon.lua:
	Added initial code for AK-47 and miniguns, and shared a lot of code
	between the weapon types.

	* basic-tilelike.lua:
	Some initial code to make crates destructible.

	* basic-powerup.lua, basic-base.lua:
	Added hide_and_respawn_later() method which is used a lot by objects.

2002-01-19  Peter Wang  <tjaden@users.sourceforge.net>

	* objgen.lua: Exposed object's id value.  This was done for corpse
	tracking after death, but I'm not too sure it should be done like
	this generally.

	* gamesrv.h, gamesrv.c, bindgen.lua: Made spawn_object,
	spawn_projectile return the objects they create.

	* gameclt.c, basic-player.lua:
	Added tracking corpses after death.

	* basic-weapon.lua:
	The start of some new weapons: shotguns and sniper rifles.

	* bindgen.lua, gamesrv.h, gamesrv.c:
	Added a delta_angle parameter to game_server_spawn_projectile().

	* object.c: Fixed collision detection so that _both_ objects'
	conditions for collision detection have to be passed before
	checking is done.

	* editor/mdobject.c, bindgen.lua:
	Allow objects without a category which don't show up in the editor.
	* blod.c: Made blod types not show up in map editor.
	* basic-player.lua: Made test corpse not show up in editor.

	* objtypes.h, objtypes.c, basic-base.lua:
	Return error code if registering an object type failed.

	* object.c:
	Slow down player if ramped movement used (walking up stairs, etc.)

	* blod.c: Added remaining blods and gave them variable masses and
	timeouts.

	* gamesrv.c: Link players at top of object list so that they
	appear above objects that were present at map load time (ladders,
	barrels, etc.)

	* gamesrv.c: Made jumping from standing on top of ladder position
	to be as high as jump from normal tile.

	* basic-player.lua: Decreased player xv_decay, yv_decay to
	slow down movement a little.

	* store/README, store/CHANGES, store/hash.c, store/store.c:
	Reverted to the libc string functions and only support
	ASCII/UTF-8 (faster).

	* basic-player.lua, bindgen.lua, gameclt.c, gameinit.c,
	* gamesrv.c, gamesrv.h, netmsg.h, blod.c, blod.h:
	Added blods.

	* object.c, object.h: Added client-processed objects.

2002-01-18  Peter Wang  <tjaden@users.sourceforge.net>

	* blood.c: Fixed a bug where it would try to allocate zero new
	particles.

	* editor/mdobject.c: Added snap to grid.

	* gamesrv.c:
	Got rid of a bad optimisation which broke down the mechanism which
	stopped players falling into ladders.

	* editor/mdobject.c:
	Added different cursors for different object editting actions.

	* editor/cursor.c:
	Created dot cursor bitmap at module initialisation instead of
	creating it everytime it was set to improve efficiency.

	* gui/guiwmdef.c, gui/gui.c:
	Added display switching in callback to force redrawing the screen.

	* editor/cursor.c:
	Created dot cursor bitmap at module initialisation instead of
	creating it every time it was set to improve efficiency.

	* editor/mdobject.c:
	Added different cursors for different object editting actions.

	* object.c:
	Use rawget, rawset instead of gettable, settable in places.


lua-4.1-work4/src/lua/lua.c
/*
** $Id: lua.c,v 1.77 2002/02/14 22:23:43 roberto Exp $
** Lua stand-alone interpreter
** See Copyright Notice in lua.h
*/


#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "lua.h"

#include "lauxlib.h"
#include "luadebug.h"
#include "lualib.h"


#ifdef _POSIX_SOURCE
#include <unistd.h>
#else
static int isatty (int x) { return x==0; }  /* assume stdin is a tty */
#endif


#ifndef LUA_PROGNAME
#define LUA_PROGNAME	"lua: "
#endif


#ifndef PROMPT
#define PROMPT		"> "
#endif


#ifndef PROMPT2
#define PROMPT2		">> "
#endif


#ifndef LUA_USERINIT
#define LUA_USERINIT(L)		openstdlibs(L)
#endif


static lua_State *L = NULL;


static lua_Hook old_linehook = NULL;
static lua_Hook old_callhook = NULL;


static void lstop (void) {
  lua_setlinehook(L, old_linehook);
  lua_setcallhook(L, old_callhook);
  lua_error(L, "interrupted!");
}


static void laction (int i) {
  (void)i;  /* to avoid warnings */
  signal(SIGINT, SIG_DFL); /* if another SIGINT happens before lstop,
                              terminate process (default action) */
  old_linehook = lua_setlinehook(L, (lua_Hook)lstop);
  old_callhook = lua_setcallhook(L, (lua_Hook)lstop);
}


/* Lua gives no message in such cases, so we provide one */
static void report (int result) {
  if (result == LUA_ERRMEM || result == LUA_ERRERR)
    fprintf(stderr, "%s%s\n", LUA_PROGNAME, luaL_errstr(result));
}


static int ldo (int (*f)(lua_State *l, const char *), const char *name,
                int clear) {
  int result;
  int top = lua_gettop(L);
  signal(SIGINT, laction);
  result = f(L, name);  /* dostring | dofile */
  signal(SIGINT, SIG_DFL);
  if (clear) lua_settop(L, top);  /* remove eventual results */
  report(result);
  return result;
}


static void print_usage (void) {
  fprintf(stderr,
  "usage: lua [options].  Available options are:\n"
  "  -        execute stdin as a file\n"
  "  -c       close Lua when exiting\n"
  "  -e stat  execute string `stat'\n"
  "  -f name  execute file `name' with remaining arguments in table `arg'\n"
  "  -i       enter interactive mode\n"
  "  -v       print version information\n"
  "  a=b      set global `a' to string `b'\n"
  "  name     execute file `name'\n"
);
}


static void print_version (void) {
  printf("%.80s  %.80s\n", LUA_VERSION, LUA_COPYRIGHT);
}


static void assign (char *arg) {
  char *eq = strchr(arg, '=');
  *eq = '\0';  /* spilt `arg' in two strings (name & value) */
  lua_pushstring(L, eq+1);
  lua_setglobal(L, arg);
}


static void getargs (char *argv[]) {
  int i;
  lua_newtable(L);
  for (i=0; argv[i]; i++) {
    /* arg[i] = argv[i] */
    lua_pushnumber(L, i);
    lua_pushstring(L, argv[i]);
    lua_rawset(L, -3);
  }
  /* arg.n = maximum index in table `arg' */
  lua_pushliteral(L, "n");
  lua_pushnumber(L, i-1);
  lua_rawset(L, -3);
}


static int l_getargs (lua_State *l) {
  char **argv = (char **)lua_touserdata(l, lua_upvalueindex(1));
  getargs(argv);
  return 1;
}


static int file_input (const char *name) {
  int result = ldo(lua_dofile, name, 1);
  if (result) {
    if (result == LUA_ERRFILE) {
      fprintf(stderr, "%s%s ", LUA_PROGNAME, luaL_errstr(result));
      perror(name);
    }
    return EXIT_FAILURE;
  }
  else
    return EXIT_SUCCESS;
}


#ifdef USE_READLINE
#include <readline/readline.h>
#include <readline/history.h>
#define save_line(b)	if (strcspn(b, " \t\n") != 0) add_history(b)
#define push_line(b)	if (incomplete) lua_pushstring(L, "\n"); lua_pushstring(L, b); free(b)
#else
#define save_line(b)
#define push_line(b)	lua_pushstring(L, b)

/* maximum length of an input line */
#ifndef MAXINPUT
#define MAXINPUT	512
#endif


static char *readline (const char *prompt) {
  static char buffer[MAXINPUT];
  if (prompt) {
    fputs(prompt, stdout);
    fflush(stdout);
  }
  return fgets(buffer, sizeof(buffer), stdin);
}
#endif


static const char *get_prompt (int incomplete) {
  const char *p = NULL;
  lua_getglobal(L, incomplete ? "_PROMPT2" : "_PROMPT");
  p = lua_tostring(L, -1);
  if (p == NULL) p = (incomplete ? PROMPT2 : PROMPT);
  lua_pop(L, 1);  /* remove global */
  return p;
}


static int incomplete = 0;

static int trap_eof (lua_State *l) {
 const char *s = lua_tostring(l, 1);
 if (strstr(s, "last token read: `<eof>'") != NULL)
   incomplete = 1;
 else
   fprintf(stderr, "error: %s\n", s);
 return 0;
}


static int load_string (void) {
  lua_getglobal(L, LUA_ERRORMESSAGE);
  lua_pushvalue(L, 1);
  lua_setglobal(L, LUA_ERRORMESSAGE);
  incomplete = 0;
  for (;;) {  /* repeat until gets a complete line */
    int result;
    char *buffer = readline(get_prompt(incomplete));
    if (buffer == NULL) {  /* input end? */
      lua_settop(L, 2);
      lua_setglobal(L, LUA_ERRORMESSAGE);
      return 0;
    }
    if (!incomplete && buffer[0] == '=') {
      buffer[0] = ' ';
      lua_pushstring(L, "return");
    }
    push_line(buffer);
    lua_concat(L, lua_gettop(L)-2);
    incomplete = 0;
    result = lua_loadbuffer(L, lua_tostring(L, 3), lua_strlen(L, 3), "=stdin");
    if (incomplete) continue;  /* repeat loop to get rest of `line' */
    save_line(lua_tostring(L, 3));
    lua_remove(L, 3);
    if (result == 0) {
      lua_insert(L, 2);  /* swap compiled chunk with old _ERRORMESSAGE */
      lua_setglobal(L, LUA_ERRORMESSAGE);  /* restore old _ERRORMESSAGE */
      return 1;
    }
    else
      report(result);
 }
}


static int lcall (lua_State *l, const char *name) {
  (void)name;  /* to avoid warnings */
  return lua_call(l, 0, LUA_MULTRET);
}


static void manual_input (int version) {
  if (version) print_version();
  lua_pushcfunction(L, trap_eof);  /* set up handler for incomplete lines */
  while (load_string()) {
    ldo(lcall, NULL, 0);
    if (lua_gettop(L) > 1) {  /* any result to print? */
      lua_getglobal(L, "print");
      lua_insert(L, 2);
      lua_call(L, lua_gettop(L)-2, 0);
    }
    else
      lua_settop(L, 1);  /* remove eventual results */
  }
  printf("\n");
  lua_settop(L, 0);  /* remove trap_eof */
}


static int handle_argv (char *argv[], int *toclose) {
  if (*argv == NULL) {  /* no more arguments? */
    if (isatty(0)) {
      manual_input(1);
    }
    else
      ldo(lua_dofile, NULL, 1);  /* executes stdin as a file */
  }
  else {  /* other arguments; loop over them */
    int i;
    for (i = 0; argv[i] != NULL; i++) {
      if (argv[i][0] != '-') {  /* not an option? */
        if (strchr(argv[i], '='))
          assign(argv[i]);
        else
          if (file_input(argv[i]) != EXIT_SUCCESS)
            return EXIT_FAILURE;  /* stop if file fails */
        }
        else switch (argv[i][1]) {  /* option */
          case '\0': {
            ldo(lua_dofile, NULL, 1);  /* executes stdin as a file */
            break;
          }
          case 'i': {
            manual_input(0);
            break;
          }
          case 'c': {
            *toclose = 1;
            break;
          }
          case 'v': {
            print_version();
            break;
          }
          case 'e': {
            i++;
            if (argv[i] == NULL) {
              print_usage();
              return EXIT_FAILURE;
            }
            if (ldo(lua_dostring, argv[i], 1) != 0) {
              fprintf(stderr, LUA_PROGNAME "error running argument `%.99s'\n",
                      argv[i]);
              return EXIT_FAILURE;
            }
            break;
          }
          case 'f': {
            i++;
            if (argv[i] == NULL) {
              print_usage();
              return EXIT_FAILURE;
            }
            getargs(argv+i);  /* collect remaining arguments */
            lua_setglobal(L, "arg");
            return file_input(argv[i]);  /* stop scanning arguments */
          }
          case 's': {
            fprintf(stderr, LUA_PROGNAME
                    "option `-s' is deprecated (dynamic stack now)\n");
            break;
          }
          default: {
            print_usage();
            return EXIT_FAILURE;
          }
        }
    }
  }
  return EXIT_SUCCESS;
}


static void register_getargs (char *argv[]) {
  lua_newuserdatabox(L, argv);
  lua_pushcclosure(L, l_getargs, 1);
  lua_setglobal(L, "getargs");
}


static void openstdlibs (lua_State *l) {
  lua_baselibopen(l);
  lua_iolibopen(l);
  lua_strlibopen(l);
  lua_mathlibopen(l);
  lua_dblibopen(l);
}


int main (int argc, char *argv[]) {
  int status;
  int toclose = 0;
  (void)argc;  /* to avoid warnings */
  L = lua_open();  /* create state */
  LUA_USERINIT(L);  /* open libraries */
  register_getargs(argv);  /* create `getargs' function */
  status = handle_argv(argv+1, &toclose);
  if (toclose)
    lua_close(L);
  return status;
}

